### PR TITLE
Remove yamlish option in yaml renderer and use as_json instead.

### PR DESCRIPTION
**Pull request by simi**

### CODE REVIEW

**Comment by sferik** [comment]
Seems good.

**Comment by simi** [comment]
I'm going to recheck 100% backward compatibility manually. But after few tests it looks same.

*I'm just wondering why YAML API is still needed.*

**Comment by sferik** [comment]
> *I'm just wondering why YAML API is still needed.*

The Ruby standard library didn’t include JSON until version 1.9.2. That means that many clients that didn’t want to specify any gem dependencies (e.g. `rubygems`) talked to RubyGems.org using YAML. The most recently results of the [Ruby survey](http://www.askr.me/ruby) show that 0.9% of respondents still use Ruby 1.8.7. If we assume that number is accurate, it means we serve around 15,000 requests to Ruby 1.8.7 clients per day.

Even if you ignore Ruby 1.8.7 users, the currently version of `rubygems` still uses that YAML API in many places (e.g. [lib/rubygems/commands/owner_command.rb](https://github.com/rubygems/rubygems/blob/4f977b3f684af97a6daa5808f9c6a7c22d318fac/lib/rubygems/commands/owner_command.rb#L59)). 

Even if we deprecate YAML today and remove it from `rubygems`, it will still a long time (at least five years) before we’ll be able to remove it without breaking Ruby.

If you really care about this issue, a good first step would be submitting a pull request to the `rubygems` client that removes all YAML requests.

**Comment by simi** [comment]
thanks @sferik for detailed explanation

**Comment by sferik** [comment]
Thanks for your patch. :smile:

**Comment by arthurnn** [comment]
Btw, seems like this option is not tested, I remember I removed it another day, and all tests were green, but the yaml output was not right.

**Comment by simi** [comment]
Actually it is. If you remove `as_json` part from renderer, a lot of API functional tests will be red. I'm going to provide 100% 1:1 test before merge to ensure.

**Comment by simi** [comment]
Actually it is. If you remove `as_json` part from renderer, a lot of API functional tests will be red. I'm going to provide 100% 1:1 test before merge to ensure.

**Comment by simi** [comment]
OK I have found one difference.

with this patch:

```yaml
--- |
  ---
  - authors: Joe User
    built_at: 2015-03-29 10:26:23.111485000 Z
    description: Some awesome gem
    downloads_count: 0
    metadata:
      foo: bar
    number: 2.0.0
    summary: 
    platform: ruby
    ruby_version: ">= 2.0.0"
    prerelease: false
    licenses: MIT
    requirements: Opencv
    sha: b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78
  - authors: Joe User
    built_at: 2015-03-29 10:26:23.111485000 Z
    description: Some awesome gem
    downloads_count: 0
    metadata:
      foo: bar
    number: 1.0.0.pre
    summary: 
    platform: ruby
    ruby_version: ">= 2.0.0"
    prerelease: true
    licenses: MIT
    requirements: Opencv
    sha: b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78
```

without:

```yaml
--- |
  ---
  - authors: Joe User
    built_at: '2015-03-29T10:26:23.111Z'
    description: Some awesome gem
    downloads_count: 0
    metadata:
      foo: bar
    number: 2.0.0
    summary: 
    platform: ruby
    ruby_version: ">= 2.0.0"
    prerelease: false
    licenses: MIT
    requirements: Opencv
    sha: b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78
  - authors: Joe User
    built_at: '2015-03-29T10:26:23.111Z'
    description: Some awesome gem
    downloads_count: 0
    metadata:
      foo: bar
    number: 1.0.0.pre
    summary: 
    platform: ruby
    ruby_version: ">= 2.0.0"
    prerelease: true
    licenses: MIT
    requirements: Opencv
    sha: b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78
```

diff:

```diff
--- without.yaml        2015-03-30 12:28:45.183434870 +0200
+++ with.yaml   2015-03-30 12:29:01.255276979 +0200
@@ -1,7 +1,7 @@
 --- |
   ---
   - authors: Joe User
-    built_at: '2015-03-29T10:26:23.111Z'
+    built_at: 2015-03-29 10:26:23.111485000 Z
     description: Some awesome gem
     downloads_count: 0
     metadata:
@@ -15,7 +15,7 @@
     requirements: Opencv
     sha: b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78
   - authors: Joe User
-    built_at: '2015-03-29T10:26:23.111Z'
+    built_at: 2015-03-29 10:26:23.111485000 Z
     description: Some awesome gem
     downloads_count: 0
     metadata:
```

That's probably backwards incompatible issue.

But I can change `data = obj.as_json.to_yaml` to `data = MultiJson.load(obj.to_json).to_yaml` to get rid of yamlish option and get the same result.

**Comment by simi** [comment]
Also I can add heavier tests (in another PR) and merge that (probably 1:1 yaml string check) before this one can be merged. What do you think? It will help rubygems.org to prevent API changes by mistake.

**Comment by sferik** [comment]
I think the unquoted (and more precise) timestamps should not be a problem. I’m pretty sure they’re not used by `rubygems` or `bundler`.

I’d :heart: to take you up on your offer to add more tests. No reason to put those in a separate pull request. Just commit them in this branch and push. Thanks!

**Comment by simi** [comment]
The idea behind separate PR was to add tests to current code and then run those tests against this patch to ensure 1:1 API compatibility for this patch and to the future

**Comment by sferik** [comment]
@simi Fair point. But as long as the only difference is the time precision, I’d be willing to merge this “breaking” change.

**Comment by arthurnn** [comment]
@simi thanks for checking all that.
I agree with @sferik . I think we can merge this, and the time precision should not be a big deal.
Actually, I plan to remove the calls we have to `MultiJSON`, so this is one step in that direction.

thanks

**Comment by qrush** [comment]
There's a lot of *what* but no *why* here - why can we use `as_json` and why do we need to switch?

**Comment by bradly** [comment]
@simi @sferik fwiw the format change is more than just precision; it goes from iso 8601 format to i think sql datetime format (not sure the name of it exactly). I'm not sure how this yaml is used to know whether that matters or not, just pointing it out.

**Comment by simi** [comment]
@qrush there are 2 reasons:

1. remove `MultiJson`
2. get rid of ugly `yamlish: true` option passed to renderer

File: config/initializers/yaml_renderer.rb [hunk 1]
```diff
@@ -1,7 +1,4 @@
-# :yamlish converts object to JSON and back before converting to YAML in order to
-# strip the object type (e.g. !ruby/ActiveRecord:Rubygem) from response
-# TODO: Remove :yamlish once we know how to strip object type with to_yaml
 ActionController::Renderers.add :yaml do |obj, options|
-  data = options[:yamlish] ? MultiJson.load(obj.to_json).to_yaml : obj.to_yaml
+  data = obj.as_json.to_yaml
   send_data data, :type => 'text/yaml'
```
Comment (arthurnn) [inline config/initializers/yaml_renderer.rb]: Wanna change this to use new hash syntax, as you are here.

Comment (simi) [inline config/initializers/yaml_renderer.rb]: done

**Comment by sferik** [comment]
@simi Can you please rebase? This no longer merged cleanly.

**Comment by simi** [comment]
@sferik done

**Comment by arthurnn** [comment]
thanks

**Pull request merged**